simulator/simulator.cpp

```diff
@@ -0,0 +1,309 @@
+//
+// Created by Павло Коваль on 16.11.2022.
+//
+
+#include "simulator.h"
+
+using namespace Info;
+
+
+Simulator::Simulator(const std::vector<std::string> &fileContent)
+{
+    for (auto &line: fileContent)
+    {
+        if (state.numMemory.to_uint() >= MAX_WORDS)
+        {
+            throw std::runtime_error("Exceed memory size");
+        }
+
+        state.mem[state.numMemory.to_uint()] = std::stoi(line);
+        state.numMemory = state.numMemory.to_uint() + 1;
+    }
+
+    for (auto &reg: state.registers)
+    {
+        reg = 0;
+    }
+}
+
+
+void Simulator::run()
+{
+    unsigned int amountOfInstructions = 0;
+    while (true)
+    {
+        opcode = state.mem[state.pc.to_uint()].to_uint() >> SHIFT_OP;
+        arg0 = (state.mem[state.pc.to_uint()].to_uint() >> SHIFT_ARG0) & 0x7;
+        arg1 = (state.mem[state.pc.to_uint()].to_uint() >> SHIFT_ARG1) & 0x7;
+        arg2 = state.mem[state.pc.to_uint()].to_uint() & 0x7;
+        addressField = state.mem[state.pc.to_uint()].to_uint() & 0xFFF;
+
+        std::cout << "addr: " << state.pc.to_uint()
+                  << "| OP: " << opcode
+                  << "| ARG0: " << arg0
+                  << "| ARG1: " << arg1
+                  << "| ARG2: " << arg2
+                  << "| Adr: " << addressField
+                  << " [" << state.mem[state.pc.to_uint()].to_uint() << "] "
+                  << std::endl;
+
+        if (opcodes[ADD] == opcode)
+        {
+            doRInstruction([](int a, int b) { return a + b; });
+        }
+        else if (opcodes[NAND] == opcode)
+        {
+            doRInstruction([](int a, int b) { return ~(a & b); });
+        }
+        else if (opcodes[LW] == opcode)
+        {
+            if (state.registers[arg0].to_int() + addressField >= MAX_WORDS)
+            {
+                throw std::runtime_error("Exceed memory size");
+            }
+
+            state.registers[arg1] = state.mem[state.registers[arg0].to_uint() + addressField];
+        }
+        else if (opcodes[SW] == opcode)
+        {
+            if (state.registers[arg0].to_uint() + addressField >= MAX_WORDS)
+            {
+                throw std::runtime_error("Exceed memory size");
+            }
+
+            state.mem[state.registers[arg0].to_uint() + addressField] = state.registers[arg1];
+        }
+        else if (opcodes[BEQ] == opcode)
+        {
+            if (state.registers[arg0].to_uint() == state.registers[arg1].to_uint())
+            {
+                state.pc = addressField-1;
+            }
+        }
+        else if (opcodes[JALR] == opcode)
+        {
+            state.registers[arg1] = state.pc;
+            if(arg0 != 0)
```

Comment:
you use 

arg0 != 0
arg0 == 0
!is_number(command.arg0)

in different scenarios.

Please use a common way of defining such a check.
Probably move it to some common place?